lib/posix/pthread_barrier.c

```diff
@@ -4,31 +4,178 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#include <zephyr/init.h>
 #include <zephyr/kernel.h>
 #include <zephyr/posix/pthread.h>
-#include <ksched.h>
-#include <zephyr/wait_q.h>
+#include <zephyr/sys/bitarray.h>
 
-extern struct k_spinlock z_pthread_spinlock;
+#include "posix_internal.h"
+
+struct posix_barrier {
+	struct k_mutex mutex;
+	struct k_condvar cond;
+	uint32_t max;
+	uint32_t count;
+};
+
+static struct posix_barrier posix_barrier_pool[CONFIG_MAX_PTHREAD_BARRIER_COUNT];
+SYS_BITARRAY_DEFINE_STATIC(posix_barrier_bitarray, CONFIG_MAX_PTHREAD_BARRIER_COUNT);
+
+/*
+ * We reserve the MSB to mark a pthread_barrier_t as initialized (from the
+ * perspective of the application). With a linear space, this means that
+ * the theoretical pthread_barrier_t range is [0,2147483647].
+ */
+BUILD_ASSERT(CONFIG_MAX_PTHREAD_BARRIER_COUNT < PTHREAD_OBJ_MASK_INIT,
+	     "CONFIG_MAX_PTHREAD_BARRIER_COUNT is too high");
+
+static inline size_t posix_barrier_to_offset(struct posix_barrier *bar)
+{
+	return bar - posix_barrier_pool;
+}
+
+static inline size_t to_posix_barrier_idx(pthread_barrier_t barrier)
+{
+	return mark_pthread_obj_uninitialized(barrier);
+}
+
+struct posix_barrier *get_posix_barrier(pthread_barrier_t b)
```

Comment:
Should this do any range checking of the value to make sure it isn't completely wild?